src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -589,6 +597,87 @@ class RNNOpRewriteLayoutPattern : public OpRewritePattern<ONNXOp> {
   }
 };
 
+// =============================================================================
+// Rewrite pattern for Power
+// =============================================================================
+
+class PowToMulRewritePattern : public OpRewritePattern<ONNXPowOp> {
+public:
+  using OpRewritePattern<ONNXPowOp>::OpRewritePattern;
+
+  PowToMulRewritePattern(MLIRContext *context, int64_t maxPower)
+      : OpRewritePattern(context), maxPower(maxPower) {}
+
+  LogicalResult matchAndRewrite(
+      ONNXPowOp powOp, PatternRewriter &rewriter) const override {
+    Operation *op = powOp.getOperation();
+    Location loc = powOp.getLoc();
+    int64_t exponent;
+    // Test legality
+    if (!CanExpandPowOpToMul(powOp, exponent))
+      return failure();
+
+    // Rewrite
+    MultiDialectBuilder<OnnxBuilder> create(rewriter, loc);
+    Value input = powOp.getX();
+
+    Value result;
+    Type resultType = powOp.getZ().getType();
+    Type elementType = getElementTypeOrSelf(resultType);
+    if (exponent == 0) {
+      DenseElementsAttr valAttr;
+      if (elementType.isa<FloatType>())
+        valAttr = DenseElementsAttr::get(resultType, ArrayRef<float>({1.0}));
+      else if (elementType.isa<IntegerType>())
+        valAttr = DenseElementsAttr::get(resultType, ArrayRef<int64_t>({1}));
+      else
+        llvm_unreachable("Unsupported type");
+      result = create.onnx.constant(valAttr);
+    } else {
+      // calculate pow(input,exponent) with "exponentiation by squaring" method
+      bool result_initialized = false;
+      while (exponent > 0) {
+        if (exponent & 1) {
+          result = result_initialized
+                       ? create.onnx.mul(resultType, result, input)
+                       : input;
+          result_initialized = true;
+        }
+        input = create.onnx.mul(resultType, input, input);
```

Comment:
would it be worthwhile avoiding the extra unused multiplication in the last loop iteration?

Reply:
@sorenlassen why is it unused? The number of `mul` must be the same as `exponent`

Reply:
I see. Good catch. Thanks for the explaination!